File: backend/ft_transcendence/http/http.py
from django.http import HttpResponse
import json


def JSONResponse(obj: dict | list, options={}) -> HttpResponse:
    """
    Generic response, avoid using it
    """

    response = {}
    response["data"] = obj
    return HttpResponse(
        json.dumps(response), content_type="application/json", **options
    )


def OK(obj: dict | list, options={}) -> HttpResponse:
    """
    https://developer.mozilla.org/pt-BR/docs/Web/HTTP/Status/200
    """
    return JSONResponse(obj, options)


def Created(obj: dict | list, options={"status": 201}) -> HttpResponse:
    """
    https://developer.mozilla.org/pt-BR/docs/Web/HTTP/Status/201
    """
    return JSONResponse(obj, options)


def NoContent(obj: dict | list, options={"status": 204}) -> HttpResponse:
    """
    https://developer.mozilla.org/pt-BR/docs/Web/HTTP/Status/204
    """
    return JSONResponse(obj, options)


def BadRequest(obj: dict | list, options={"status": 400}) -> HttpResponse:
    """
    https://developer.mozilla.org/pt-BR/docs/Web/HTTP/Status/400
    """
    return JSONResponse(obj, options)


def Unauthorized(obj: dict | list, options={"status": 401}) -> HttpResponse:
    """
    https://developer.mozilla.org/pt-BR/docs/Web/HTTP/Status/401
    """
    return JSONResponse(obj, options)


def Forbidden(obj: dict | list, options={"status": 403}) -> HttpResponse:
    """
    https://developer.mozilla.org/pt-BR/docs/Web/HTTP/Status/403
    """
    return JSONResponse(obj, options)


def NotFound(obj: dict | list, options={"status": 404}) -> HttpResponse:
    """
    https://developer.mozilla.org/pt-BR/docs/Web/HTTP/Status/404
    """
    return JSONResponse(obj, options)


def MethodNotAllowed(options={"status": 405}) -> HttpResponse:
    """
    https://developer.mozilla.org/pt-BR/docs/Web/HTTP/Status/405
    """
    return JSONResponse({"message": "Method not allowed"}, options)


def TooManyRequests(obj: dict | list, options={"status": 429}) -> HttpResponse:
    """
    https://developer.mozilla.org/pt-BR/docs/Web/HTTP/Status/429
    """
    return JSONResponse(obj, options)


def InternalServerError(obj: dict | list, options={"status": 500}) -> HttpResponse:
    """
    https://developer.mozilla.org/pt-BR/docs/Web/HTTP/Status/500
    """
    return JSONResponse(obj, options)

File: backend/static/js/services/http.mjs
export const HTTPStatus = Object.freeze({
  OK: 200,
  CREATED: 201,
  NO_CONTENT: 204,
  BAD_REQUEST: 400,
  UNAUTHORIZED: 401,
  FORBIDDEN: 403,
  NOT_FOUND: 404,
  METHOD_NOT_ALLOWED: 405,
  UNPROCESSABLE_CONTENT: 422,
  TOO_MANY_REQUESTS: 429,
});

export class RequestFailedError extends Error {
  /** @type {number} */
  status;
  /** @type {Response} */
  response;
  /** @type {Record<string, any>} */
  error;

  /**
   * @param {Response} response
   */
  constructor(response) {
    super();
    this.response = response;
    this.status = response.status;
    this.error = response.json();
  }
}

/**
 * @param {string} path
 * @param {Exclude<Parameters<typeof fetch>[0], URL>} options
 */
export async function http(path, options) {
  const h = new Headers();

  if (options.headers) {
    for (const key in options.headers) {
      h.append(key, options.headers[key]);
    }
  }

  const response = await fetch(path, {
    ...options,
    headers: h,
    credentials: "same-origin",
  });

  if (response.status >= HTTPStatus.BAD_REQUEST) {
    throw new RequestFailedError(response);
  }

  try {
    const data = await response.json();
    return {
      data,
      response,
    };
  } catch (e) {
    console.error(e);
    return { data: { data: { message: "Error parsing JSON" } }, response };
  }
}

/**
 * @param {string} path
 * @param {Exclude<Parameters<typeof fetch>[0], URL>} options
 */
export async function GET(path, options) {
  return http(path, {
    ...options,
    method: "GET",
  });
}

/**
 * @param {string} path
 * @param {object} payload
 * @param {Exclude<Parameters<typeof fetch>[0], URL>} options
 */
export async function POST(path, payload, options) {
  return http(path, {
    ...options,
    body: JSON.stringify(payload),
    method: "POST",
  });
}

/**
 * @param {string} path
 * @param {Exclude<Parameters<typeof fetch>[0], URL>} options
 */
export async function PUT(path, options) {
  return http(path, {
    ...options,
    method: "PUT",
  });
}

/**
 * @param {string} path
 * @param {Exclude<Parameters<typeof fetch>[0], URL>} options
 */
export async function DELETE(path, options) {
  return http(path, {
    ...options,
    method: "DELETE",
  });
}
